File: Image-Resizing-OpenCV/resize_image.cpp
#include <algorithm>
#include <cmath>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

namespace fs = std::filesystem;

struct LetterboxResult {
    cv::Mat image;
    double scale;
    cv::Rect content;
};

cv::Mat resizeExact(const cv::Mat& image, int width, int height, int interpolation) {
    if (image.empty()) throw std::invalid_argument("image must be non-empty");
    if (width <= 0 || height <= 0) {
        throw std::invalid_argument("target width and height must be positive");
    }
    cv::Mat output;
    cv::resize(image, output, cv::Size(width, height), 0.0, 0.0, interpolation);
    return output;
}

cv::Mat resizeByScale(const cv::Mat& image, double scaleX, double scaleY,
                      int interpolation) {
    if (image.empty()) throw std::invalid_argument("image must be non-empty");
    if (scaleX <= 0.0 || scaleY <= 0.0) {
        throw std::invalid_argument("scale factors must be positive");
    }
    cv::Mat output;
    cv::resize(image, output, cv::Size(), scaleX, scaleY, interpolation);
    return output;
}

LetterboxResult letterbox(const cv::Mat& image, int targetWidth, int targetHeight,
                          const cv::Scalar& color = cv::Scalar(32, 32, 32)) {
    if (image.empty()) throw std::invalid_argument("image must be non-empty");
    if (targetWidth <= 0 || targetHeight <= 0) {
        throw std::invalid_argument("target width and height must be positive");
    }
    const double scale = std::min(
        static_cast<double>(targetWidth) / image.cols,
        static_cast<double>(targetHeight) / image.rows);
    const int contentWidth = std::max(1, static_cast<int>(std::lround(image.cols * scale)));
    const int contentHeight = std::max(1, static_cast<int>(std::lround(image.rows * scale)));
    cv::Mat resized = resizeExact(
        image, contentWidth, contentHeight,
        scale < 1.0 ? cv::INTER_AREA : cv::INTER_CUBIC);
    cv::Mat output(targetHeight, targetWidth, image.type(), color);
    const cv::Rect content((targetWidth - contentWidth) / 2,
                           (targetHeight - contentHeight) / 2,
                           contentWidth, contentHeight);
    resized.copyTo(output(content));
    return {output, scale, content};
}

cv::Mat labeledPanel(const cv::Mat& image, const std::string& label) {
    cv::Mat panel = image.clone();
    cv::rectangle(panel, cv::Rect(0, 0, panel.cols, 42), cv::Scalar::all(0), -1);
    cv::putText(panel, label, cv::Point(14, 29), cv::FONT_HERSHEY_SIMPLEX,
                0.72, cv::Scalar::all(255), 2, cv::LINE_AA);
    return panel;
}

cv::Mat makeComparison(const cv::Mat& image) {
    const cv::Size half(image.cols / 2, image.rows / 2);
    cv::Mat areaSmall;
    cv::Mat linearSmall;
    cv::resize(image, areaSmall, half, 0.0, 0.0, cv::INTER_AREA);
    cv::resize(image, linearSmall, half, 0.0, 0.0, cv::INTER_LINEAR);
    cv::Mat areaRestored;
    cv::Mat linearRestored;
    cv::resize(areaSmall, areaRestored, image.size(), 0.0, 0.0, cv::INTER_NEAREST);
    cv::resize(linearSmall, linearRestored, image.size(), 0.0, 0.0, cv::INTER_NEAREST);
    std::vector<cv::Mat> panels{
        labeledPanel(image, "Input"),
        labeledPanel(areaRestored, "Downscale: INTER_AREA"),
        labeledPanel(linearRestored, "Downscale: INTER_LINEAR")};
    cv::Mat comparison;
    cv::hconcat(panels, comparison);
    return comparison;
}

void requireWrite(const fs::path& path, const cv::Mat& image) {
    fs::create_directories(path.parent_path());
    if (!cv::imwrite(path.string(), image)) {
        throw std::runtime_error("could not write " + path.string());
    }
}

int selfTest() {
    cv::Mat fixture(40, 80, CV_8UC3, cv::Scalar(10, 20, 30));
    const cv::Mat exact = resizeExact(fixture, 20, 10, cv::INTER_AREA);
    const cv::Mat scaled = resizeByScale(fixture, 0.5, 0.5, cv::INTER_AREA);
    const LetterboxResult boxed = letterbox(fixture, 100, 100);
    const cv::Mat oddGeometry(4, 5, CV_8UC3, cv::Scalar::all(0));
    const LetterboxResult oddBoxed = letterbox(oddGeometry, 4, 2);
    bool rejectedEmpty = false;
    try {
        static_cast<void>(letterbox(cv::Mat(), 100, 100));
    } catch (const std::invalid_argument&) {
        rejectedEmpty = true;
    }
    if (exact.size() != cv::Size(20, 10) ||
        scaled.size() != cv::Size(40, 20) ||
        boxed.image.size() != cv::Size(100, 100) ||
        boxed.content.size() != cv::Size(100, 50) ||
        boxed.content.y != 25 ||
        oddBoxed.content.size() != cv::Size(3, 2) ||
        !rejectedEmpty) {
        std::cerr << "self-test failed\n";
        return 1;
    }
    std::cout << "self-test passed\n";
    return 0;
}

int main(int argc, char** argv) {
    try {
        fs::path input = fs::path(__FILE__).parent_path() / "assets" / "sample-scene.png";
        fs::path outputDir = fs::path(__FILE__).parent_path() / "outputs";
        bool display = false;
        for (int i = 1; i < argc; ++i) {
            const std::string arg = argv[i];
            if (arg == "--self-test") return selfTest();
            if (arg == "--display") display = true;
            else if (arg == "--input" && i + 1 < argc) input = argv[++i];
            else if (arg == "--output-dir" && i + 1 < argc) outputDir = argv[++i];
            else if (arg != "--display") throw std::invalid_argument("unknown or incomplete option: " + arg);
        }
        const cv::Mat image = cv::imread(input.string(), cv::IMREAD_COLOR);
        if (image.empty()) throw std::runtime_error("could not read " + input.string());
        const cv::Mat downArea = resizeExact(
            image, image.cols / 2, image.rows / 2, cv::INTER_AREA);
        const cv::Mat upLinear = resizeByScale(image, 1.5, 1.5, cv::INTER_LINEAR);
        const cv::Mat upCubic = resizeByScale(image, 1.5, 1.5, cv::INTER_CUBIC);
        const LetterboxResult boxed = letterbox(image, 640, 640);
        requireWrite(outputDir / "downscale-inter-area.png", downArea);
        requireWrite(outputDir / "upscale-inter-linear.png", upLinear);
        requireWrite(outputDir / "upscale-inter-cubic.png", upCubic);
        requireWrite(outputDir / "letterbox-640.png", boxed.image);
        const cv::Mat comparison = makeComparison(image);
        requireWrite(outputDir / "resize-comparison.png", comparison);
        std::cout << "input=" << image.cols << "x" << image.rows
                  << " down=" << downArea.cols << "x" << downArea.rows
                  << " letterbox_content=" << boxed.content.width << "x"
                  << boxed.content.height << '\n';
        if (display) {
            cv::imshow("Resize interpolation comparison", comparison);
            cv::waitKey(0);
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 1;
    }
}

File: Image-Resizing-OpenCV/resize_image.py
#!/usr/bin/env python3
"""Resize images with explicit geometry and interpolation choices."""

from __future__ import annotations

import argparse
from dataclasses import dataclass
from pathlib import Path

import cv2
import numpy as np


PROJECT_DIR = Path(__file__).resolve().parent
DEFAULT_INPUT = PROJECT_DIR / "assets" / "sample-scene.png"


@dataclass(frozen=True)
class LetterboxResult:
    image: np.ndarray
    scale: float
    offset_x: int
    offset_y: int
    content_width: int
    content_height: int


def read_bgr(path: Path) -> np.ndarray:
    image = cv2.imread(str(path), cv2.IMREAD_COLOR)
    if image is None:
        raise FileNotFoundError(f"Could not read input image: {path}")
    return image


def resize_exact(
    image: np.ndarray, width: int, height: int, interpolation: int
) -> np.ndarray:
    if image.size == 0:
        raise ValueError("image must be non-empty")
    if width <= 0 or height <= 0:
        raise ValueError("target width and height must be positive")
    return cv2.resize(image, (width, height), interpolation=interpolation)


def resize_by_scale(
    image: np.ndarray, scale_x: float, scale_y: float, interpolation: int
) -> np.ndarray:
    if image.size == 0:
        raise ValueError("image must be non-empty")
    if scale_x <= 0 or scale_y <= 0:
        raise ValueError("scale factors must be positive")
    return cv2.resize(
        image, None, fx=scale_x, fy=scale_y, interpolation=interpolation
    )


def rounded_positive_size(value: float) -> int:
    """Round a positive size half up, matching C++ std::lround."""
    return max(1, int(value + 0.5))


def resize_to_fit(
    image: np.ndarray,
    max_width: int,
    max_height: int,
    allow_upscale: bool = False,
) -> tuple[np.ndarray, float]:
    if image.size == 0:
        raise ValueError("image must be non-empty")
    if max_width <= 0 or max_height <= 0:
        raise ValueError("bounding-box dimensions must be positive")
    height, width = image.shape[:2]
    scale = min(max_width / width, max_height / height)
    if not allow_upscale:
        scale = min(scale, 1.0)
    target = (
        rounded_positive_size(width * scale),
        rounded_positive_size(height * scale),
    )
    interpolation = cv2.INTER_AREA if scale < 1.0 else cv2.INTER_CUBIC
    return cv2.resize(image, target, interpolation=interpolation), scale


def letterbox(
    image: np.ndarray,
    target_width: int,
    target_height: int,
    color: tuple[int, int, int] = (32, 32, 32),
) -> LetterboxResult:
    resized, scale = resize_to_fit(
        image, target_width, target_height, allow_upscale=True
    )
    output = np.full((target_height, target_width, 3), color, dtype=image.dtype)
    offset_x = (target_width - resized.shape[1]) // 2
    offset_y = (target_height - resized.shape[0]) // 2
    output[
        offset_y : offset_y + resized.shape[0],
        offset_x : offset_x + resized.shape[1],
    ] = resized
    return LetterboxResult(
        image=output,
        scale=scale,
        offset_x=offset_x,
        offset_y=offset_y,
        content_width=resized.shape[1],
        content_height=resized.shape[0],
    )


def labeled_panel(image: np.ndarray, label: str) -> np.ndarray:
    panel = image.copy()
    cv2.rectangle(panel, (0, 0), (panel.shape[1], 42), (0, 0, 0), -1)
    cv2.putText(
        panel,
        label,
        (14, 29),
        cv2.FONT_HERSHEY_SIMPLEX,
        0.72,
        (255, 255, 255),
        2,
        cv2.LINE_AA,
    )
    return panel


def make_comparison(image: np.ndarray) -> np.ndarray:
    size = (image.shape[1] // 2, image.shape[0] // 2)
    area_small = cv2.resize(image, size, interpolation=cv2.INTER_AREA)
    linear_small = cv2.resize(image, size, interpolation=cv2.INTER_LINEAR)
    area_restored = cv2.resize(
        area_small, (image.shape[1], image.shape[0]), interpolation=cv2.INTER_NEAREST
    )
    linear_restored = cv2.resize(
        linear_small, (image.shape[1], image.shape[0]), interpolation=cv2.INTER_NEAREST
    )
    return cv2.hconcat(
        [
            labeled_panel(image, "Input"),
            labeled_panel(area_restored, "Downscale: INTER_AREA"),
            labeled_panel(linear_restored, "Downscale: INTER_LINEAR"),
        ]
    )


def run_resize_demo(input_path: Path, output_dir: Path) -> dict[str, float | int]:
    image = read_bgr(input_path)
    down_area = resize_exact(
        image, image.shape[1] // 2, image.shape[0] // 2, cv2.INTER_AREA
    )
    up_linear = resize_by_scale(image, 1.5, 1.5, cv2.INTER_LINEAR)
    up_cubic = resize_by_scale(image, 1.5, 1.5, cv2.INTER_CUBIC)
    boxed = letterbox(image, 640, 640)
    outputs = {
        output_dir / "downscale-inter-area.png": down_area,
        output_dir / "upscale-inter-linear.png": up_linear,
        output_dir / "upscale-inter-cubic.png": up_cubic,
        output_dir / "letterbox-640.png": boxed.image,
        output_dir / "resize-comparison.png": make_comparison(image),
    }
    output_dir.mkdir(parents=True, exist_ok=True)
    for path, output in outputs.items():
        if not cv2.imwrite(str(path), output):
            raise RuntimeError(f"Could not write output image: {path}")
    return {
        "input_width": image.shape[1],
        "input_height": image.shape[0],
        "down_width": down_area.shape[1],
        "down_height": down_area.shape[0],
        "letterbox_scale": boxed.scale,
        "letterbox_content_width": boxed.content_width,
        "letterbox_content_height": boxed.content_height,
    }


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", type=Path, default=DEFAULT_INPUT)
    parser.add_argument("--output-dir", type=Path, default=PROJECT_DIR / "outputs")
    parser.add_argument("--display", action="store_true")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    metrics = run_resize_demo(args.input, args.output_dir)
    print(
        f"input={metrics['input_width']}x{metrics['input_height']} "
        f"down={metrics['down_width']}x{metrics['down_height']} "
        f"letterbox_content={metrics['letterbox_content_width']}x"
        f"{metrics['letterbox_content_height']}"
    )
    if args.display:
        cv2.imshow("Resize interpolation comparison", make_comparison(read_bgr(args.input)))
        cv2.waitKey(0)
        cv2.destroyAllWindows()


if __name__ == "__main__":
    main()
